test/e2e/crd_versions_with_gk_scoping_test.go

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	uitest "github.com/cppforlife/go-cli-ui/ui/test"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCRDVersionsWithGKScoping(t *testing.T) {
+	env := BuildEnv(t)
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, env.KappBinaryPath, logger}
+
+	crdYamlTemplate := `
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: foostores.demo.com
+spec:
+  group: demo.com
+  versions:
+    - name: v1beta1
+      served: %s
+      storage: false
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                foo:
+                  type: string
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                foo:
+                  type: string
+  scope: Namespaced
+  names:
+    plural: foostores
+    singular: foostore
+    kind: FooStore
+    shortNames:
+    - fst	
+`
+
+	crdWithV1beta1Served := fmt.Sprintf(crdYamlTemplate, "true")
+
+	crYaml := `
+---
+apiVersion: demo.com/v1beta1
+kind: FooStore
+metadata:
+  name: test-cr
+spec:
+  foo: bar1
+`
+	name := "test-crd-versions-with-gk-scoping"
+
+	// CRD with versions v1beta1 and v1 served
+	// CR using version v1beta1
+	yaml1 := fmt.Sprintf("%s\n%s", crdWithV1beta1Served, crYaml)
+
+	// Updating CRD to stop serving version v1beta1
+	// Removing CR from the manifest
+	yaml2 := fmt.Sprintf(crdYamlTemplate, "false")
+
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", name})
+	}
+	defer cleanUp()
+
+	logger.Section("initial deploy", func() {
+		out, _ := kapp.RunWithOpts([]string{"deploy", "-a", name, "-f", "-", "--json"}, RunOpts{IntoNs: true, StdinReader: strings.NewReader(yaml1)})
+
+		resp := uitest.JSONUIFromBytes(t, []byte(out))
+		respRows := resp.Tables[0].Rows
+
+		expected := []map[string]string{
+			{
+				"age":             "",
+				"conditions":      "",
+				"kind":            "CustomResourceDefinition",
+				"name":            "foostores.demo.com",
+				"namespace":       "(cluster)",
+				"op":              "create",
+				"op_strategy":     "",
+				"reconcile_info":  "",
+				"reconcile_state": "",
+				"wait_to":         "reconcile",
+			},
+			{
+				"age":             "",
+				"conditions":      "",
+				"kind":            "FooStore",
+				"name":            "test-cr",
+				"namespace":       "kapp-test",
+				"op":              "create",
+				"op_strategy":     "",
+				"reconcile_info":  "",
+				"reconcile_state": "",
+				"wait_to":         "reconcile",
+			},
+		}
+
+		require.Exactlyf(t, expected, respRows, "Expected to see correct changes")
+	})
+
+	logger.Section("stop serving CRD on v1beta1 and remove cr from manifest", func() {
```

Comment:
Should we first deploy the crd with the changes and then do another deployment with the CR being removed?

Reply:
We could definitely do that. I went with this because ordering should kick in and have the same affect.
You think doing it that way will make the case more legible?

Reply:
I think this test would work for GVK scoping as well. But if we delete the crd in the next step, then the results for GK scoping and GVK scoping will be different.

Reply:
In case of GVK scoping.
There won't be a delete as we won't list v1 resources.
Could you elaborate ? Not sure if I am missing something here

Reply:
> we won't be listing v1 resources

We won't have to. kapp would list the v1beta1 and show the summary based on that, only once the crd is updated, the v1beta1 version would cease to exists.

Reply:
Oh yes, I see it now, makes complete sense. Will make the change

Reply:
Made tests stricter 🚀